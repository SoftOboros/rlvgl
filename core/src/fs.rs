//! Asset loading interfaces for filesystem-backed content.
//!
//! This module provides traits used by the optional `fs` feature to source
//! assets such as fonts or images from an underlying filesystem.

use alloc::boxed::Box;

/// Errors that can occur during filesystem operations.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FsError {
    /// Underlying device reported an error.
    Device,
    /// Provided path was invalid.
    InvalidPath,
    /// File or directory was not found.
    NoSuchFile,
}

/// Block device abstraction used by the filesystem layer.
///
/// Implementors provide sector-based access to a storage medium.
pub trait BlockDevice {
    /// Read blocks starting at `lba` into `buf`.
    fn read_blocks(&mut self, lba: u64, buf: &mut [u8]) -> Result<(), FsError>;

    /// Write blocks starting at `lba` from `buf`.
    ///
    /// Implementations may leave this unimplemented if the device is read-only.
    fn write_blocks(&mut self, lba: u64, buf: &[u8]) -> Result<(), FsError>;

    /// Return the logical block size in bytes.
    fn block_size(&self) -> usize;

    /// Return the total number of addressable blocks.
    fn num_blocks(&self) -> u64;

    /// Flush any buffered data to the underlying device.
    fn flush(&mut self) -> Result<(), FsError>;
}

/// Error type returned by asset operations.
#[derive(Debug, Clone)]
pub enum AssetError {
    /// Underlying filesystem error.
    Fs(FsError),
}

/// Reader trait for streaming asset data.
pub trait AssetRead {
    /// Read data into `out`, returning the number of bytes read.
    fn read(&mut self, out: &mut [u8]) -> Result<usize, AssetError>;

    /// Total length of the asset in bytes.
    fn len(&self) -> usize;

    /// Return `true` if the asset has a length of zero bytes.
    fn is_empty(&self) -> bool;

    /// Seek to an absolute byte position within the asset.
    fn seek(&mut self, pos: u64) -> Result<u64, AssetError>;
}

/// Source of assets such as fonts or images.
pub trait AssetSource {
    /// Open an asset by logical path, e.g., `"fonts/regular.bin"`.
    fn open<'a>(&'a self, path: &str) -> Result<Box<dyn AssetRead + 'a>, AssetError>;

    /// Determine whether an asset at `path` exists.
    fn exists(&self, path: &str) -> bool;

    /// List the contents of `dir`, returning an iterator over asset entries.
    fn list(&self, dir: &str) -> Result<AssetIter, AssetError>;
}

/// Iterator over asset entries returned by [`AssetSource::list`].
pub struct AssetIter;

impl Iterator for AssetIter {
    type Item = (); // placeholder until fleshed out

    fn next(&mut self) -> Option<Self::Item> {
        None
    }
}

/// Manager that provides convenient typed loading helpers.
pub struct AssetManager<S: AssetSource> {
    source: S,
}

impl<S: AssetSource> AssetManager<S> {
    /// Create a new [`AssetManager`] from an [`AssetSource`].
    pub fn new(source: S) -> Self {
        Self { source }
    }

    /// Open a raw asset stream at `path`.
    pub fn open(&self, path: &str) -> Result<Box<dyn AssetRead + '_>, AssetError> {
        self.source.open(path)
    }
}
