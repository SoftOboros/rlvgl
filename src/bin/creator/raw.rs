//! Raw RGBA image sequence format utilities for rlvgl-creator.
//!
//! Provides a simple `.raw` container with a max-frame header and optional
//! per-frame headers. Single images omit frame headers and store pixel data
//! directly after the file header.

#![allow(dead_code)]

use std::collections::HashMap;
use std::fs::File;
use std::io::{Read, Write};
use std::path::Path;

use anyhow::{Result, anyhow};
use image::{DynamicImage, GenericImageView};

/// Magic identifier for `.raw` files.
const MAGIC: &[u8; 8] = b"RLVGLRAW";

/// A single frame within a raw image sequence.
#[derive(Debug)]
pub struct Frame {
    /// X offset of the frame within the maximum bounds.
    pub x: u32,
    /// Y offset of the frame within the maximum bounds.
    pub y: u32,
    /// Width of the frame in pixels.
    pub width: u32,
    /// Height of the frame in pixels.
    pub height: u32,
    /// Raw RGBA8 pixel data for the frame.
    pub data: Vec<u8>,
}

/// Raw image sequence container.
#[derive(Debug)]
pub struct Sequence {
    /// Maximum width of any frame.
    pub max_width: u32,
    /// Maximum height of any frame.
    pub max_height: u32,
    /// Frames contained in the sequence.
    pub frames: Vec<Frame>,
}

impl Sequence {
    /// Create a sequence from a single [`DynamicImage`].
    pub fn from_image(img: DynamicImage) -> Self {
        let (width, height) = img.dimensions();
        Sequence {
            max_width: width,
            max_height: height,
            frames: vec![Frame {
                x: 0,
                y: 0,
                width,
                height,
                data: img.to_rgba8().into_raw(),
            }],
        }
    }

    /// Encode the sequence to a `.raw` file, applying simple token-based RLE
    /// compression when beneficial.
    pub fn encode<P: AsRef<Path>>(&self, path: P) -> Result<()> {
        let mut file = File::create(path)?;
        file.write_all(MAGIC)?;
        file.write_all(&self.max_width.to_le_bytes())?;
        file.write_all(&self.max_height.to_le_bytes())?;
        file.write_all(&(self.frames.len() as u32).to_le_bytes())?;

        if self.frames.len() == 1 {
            encode_frame(&mut file, &self.frames[0], true)?;
        } else {
            for f in &self.frames {
                file.write_all(&f.x.to_le_bytes())?;
                file.write_all(&f.y.to_le_bytes())?;
                file.write_all(&f.width.to_le_bytes())?;
                file.write_all(&f.height.to_le_bytes())?;
                encode_frame(&mut file, f, false)?;
            }
        }
        Ok(())
    }

    /// Decode a sequence from a `.raw` file.
    pub fn decode<P: AsRef<Path>>(path: P) -> Result<Self> {
        let mut file = File::open(path)?;
        let mut magic = [0u8; 8];
        file.read_exact(&mut magic)?;
        if &magic != MAGIC {
            return Err(anyhow!("invalid raw magic"));
        }

        let mut buf = [0u8; 4];
        file.read_exact(&mut buf)?;
        let max_width = u32::from_le_bytes(buf);
        file.read_exact(&mut buf)?;
        let max_height = u32::from_le_bytes(buf);
        file.read_exact(&mut buf)?;
        let frame_count = u32::from_le_bytes(buf);

        let mut frames = Vec::with_capacity(frame_count as usize);
        if frame_count == 1 {
            frames.push(decode_frame(&mut file, max_width, max_height, true)?);
        } else {
            for _ in 0..frame_count {
                file.read_exact(&mut buf)?;
                let x = u32::from_le_bytes(buf);
                file.read_exact(&mut buf)?;
                let y = u32::from_le_bytes(buf);
                file.read_exact(&mut buf)?;
                let width = u32::from_le_bytes(buf);
                file.read_exact(&mut buf)?;
                let height = u32::from_le_bytes(buf);
                frames.push(decode_frame(&mut file, width, height, false)?.with_pos(x, y));
            }
        }

        Ok(Sequence {
            max_width,
            max_height,
            frames,
        })
    }
}

/// Convert a `DynamicImage` directly to a `.raw` file.
pub fn encode_image<P: AsRef<Path>>(img: DynamicImage, path: P) -> Result<()> {
    Sequence::from_image(img).encode(path)
}

/// Encode a frame to the file, optionally omitting position/size data.
fn encode_frame(file: &mut File, frame: &Frame, is_single: bool) -> Result<()> {
    if !is_single {
        // position and size already written by caller
    }
    let (tokens, indices) = tokenize(&frame.data);
    if tokens.is_empty() || tokens.len() > 256 {
        file.write_all(&0u32.to_le_bytes())?;
        file.write_all(&frame.data)?;
        return Ok(());
    }

    file.write_all(&(tokens.len() as u32).to_le_bytes())?;
    for t in &tokens {
        file.write_all(t)?;
    }
    let compressed = rle_encode(&indices);
    file.write_all(&(compressed.len() as u32).to_le_bytes())?;
    file.write_all(&compressed)?;
    Ok(())
}

/// Decode a frame from the file, using the given dimensions.
fn decode_frame(file: &mut File, width: u32, height: u32, is_single: bool) -> Result<Frame> {
    if !is_single {
        // position handled by caller
    }
    let mut buf = [0u8; 4];
    file.read_exact(&mut buf)?;
    let token_count = u32::from_le_bytes(buf);
    if token_count == 0 {
        let mut data = vec![0u8; (width * height * 4) as usize];
        file.read_exact(&mut data)?;
        return Ok(Frame {
            x: 0,
            y: 0,
            width,
            height,
            data,
        });
    }

    let mut tokens = Vec::with_capacity(token_count as usize);
    for _ in 0..token_count {
        let mut t = [0u8; 4];
        file.read_exact(&mut t)?;
        tokens.push(t);
    }
    file.read_exact(&mut buf)?;
    let comp_len = u32::from_le_bytes(buf);
    let mut comp = vec![0u8; comp_len as usize];
    file.read_exact(&mut comp)?;
    let indices = rle_decode(&comp);
    let mut data = Vec::with_capacity(indices.len() * 4);
    for idx in indices {
        data.extend_from_slice(&tokens[idx as usize]);
    }
    Ok(Frame {
        x: 0,
        y: 0,
        width,
        height,
        data,
    })
}

/// Build a token table and index stream for the pixel data.
fn tokenize(data: &[u8]) -> (Vec<[u8; 4]>, Vec<u8>) {
    let mut map: HashMap<u32, u8> = HashMap::new();
    let mut tokens = Vec::new();
    let mut indices = Vec::new();

    for chunk in data.chunks_exact(4) {
        let key = u32::from_le_bytes([chunk[0], chunk[1], chunk[2], chunk[3]]);
        if let Some(&idx) = map.get(&key) {
            indices.push(idx);
        } else {
            let idx = tokens.len() as u8;
            if idx == u8::MAX {
                return (Vec::new(), Vec::new());
            }
            tokens.push([chunk[0], chunk[1], chunk[2], chunk[3]]);
            map.insert(key, idx);
            indices.push(idx);
        }
    }

    (tokens, indices)
}

/// Run-length encode a slice of token indices.
fn rle_encode(indices: &[u8]) -> Vec<u8> {
    let mut out = Vec::new();
    if indices.is_empty() {
        return out;
    }
    let mut prev = indices[0];
    let mut run: u8 = 1;
    for &b in &indices[1..] {
        if b == prev && run < u8::MAX {
            run += 1;
        } else {
            out.push(run);
            out.push(prev);
            prev = b;
            run = 1;
        }
    }
    out.push(run);
    out.push(prev);
    out
}

/// Decode a run-length encoded slice of token indices.
fn rle_decode(data: &[u8]) -> Vec<u8> {
    let mut out = Vec::new();
    let mut iter = data.chunks_exact(2);
    for pair in &mut iter {
        let run = pair[0];
        let val = pair[1];
        out.extend(std::iter::repeat(val).take(run as usize));
    }
    out
}

impl Frame {
    /// Attach position information after decoding.
    fn with_pos(mut self, x: u32, y: u32) -> Self {
        self.x = x;
        self.y = y;
        self
    }
}
