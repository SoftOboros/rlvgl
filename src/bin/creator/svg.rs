//! SVG command for rlvgl-creator.
//!
//! Renders an SVG file into one or more raw RGBA images at specific DPI values,
//! optionally applying a monochrome threshold suitable for e-ink displays.

use std::fs;
use std::path::Path;

use anyhow::{Result, anyhow};
use resvg::render;
use resvg::tiny_skia;
use resvg::usvg::{Options, Tree};

use crate::raw;

/// Render an SVG into raw RGBA images.
pub(crate) fn run(svg: &Path, out: &Path, dpis: &[f32], threshold: Option<u8>) -> Result<()> {
    let data = fs::read(svg)?;
    let name = svg
        .file_stem()
        .and_then(|s| s.to_str())
        .ok_or_else(|| anyhow!("invalid SVG name"))?;

    for dpi in dpis {
        let mut opt = Options::default();
        opt.dpi = *dpi;
        let tree = Tree::from_data(&data, &opt)?;
        let size = tree.size().to_int_size();
        let mut pixmap = tiny_skia::Pixmap::new(size.width(), size.height())
            .ok_or_else(|| anyhow!("invalid SVG size"))?;
        render(&tree, tiny_skia::Transform::default(), &mut pixmap.as_mut());
        let mut img =
            image::RgbaImage::from_raw(size.width(), size.height(), pixmap.data().to_vec())
                .ok_or_else(|| anyhow!("bad pixmap data"))?;

        if let Some(th) = threshold {
            for p in img.pixels_mut() {
                let gray = (0.299 * p[0] as f32 + 0.587 * p[1] as f32 + 0.114 * p[2] as f32) as u8;
                let v = if gray > th { 0xFF } else { 0x00 };
                *p = image::Rgba([v, v, v, p[3]]);
            }
        }

        fs::create_dir_all(out)?;
        let out_path = out.join(format!("{name}_{dpi}dpi.raw"));
        raw::encode_image(image::DynamicImage::ImageRgba8(img), &out_path)?;
        println!(
            "Rendered {} at {} DPI -> {}",
            svg.display(),
            dpi,
            out_path.display()
        );
    }

    Ok(())
}

#[cfg(all(test, feature = "regression"))]
mod tests {
    use super::*;
    use blake3::hash;
    use tempfile::tempdir;

    #[test]
    fn svg_renders_with_stable_hash() {
        let dir = tempdir().unwrap();
        let svg_path = dir.path().join("test.svg");
        fs::write(
            &svg_path,
            "<svg width=\"2\" height=\"2\" xmlns=\"http://www.w3.org/2000/svg\"><rect width=\"2\" height=\"2\" fill=\"#010203\"/></svg>",
        )
        .unwrap();
        let out_dir = dir.path().join("out");
        run(&svg_path, &out_dir, &[96.0], None).unwrap();
        let data = fs::read(out_dir.join("test_96dpi.raw")).unwrap();
        let hash = hash(&data).to_hex();
        insta::assert_snapshot!("svg_hashes", hash);
    }
}
