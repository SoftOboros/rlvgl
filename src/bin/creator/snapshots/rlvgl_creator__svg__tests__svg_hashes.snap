---
source: src/bin/creator/svg.rs
expression: "hash"
---
4afd9d2454855dc8557412527aa5a10170883fb8773f2127cb9929c51b0d51b4
